### PR TITLE
const thread-safe packed classes

### PR DESCRIPTION
**Pull request by Dr15Jones**
Made the mutable members of pat::PackedCandidate and pat::PackedGenParticle being changed in const functions be thread-safe. Added unit tests.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
I didn't know of anyway to actually test that reading back the classes work correctly.

**Comment by Dr15Jones** [comment]
please test

File: DataFormats/PatCandidates/interface/PackedGenParticle.h [hunk 1]
```diff
@@ -30,16 +34,65 @@ namespace pat {
...
-  explicit PackedGenParticle( const reco::GenParticle & c)
-    : p4_(c.pt(), c.eta(), c.phi(), c.mass()), p4c_(p4_), vertex_(0,0,0), pdgId_(c.pdgId()), charge_(c.charge()), mother_(c.motherRef(0)), unpacked_(true),
+    PackedGenParticle()
+    : packedPt_(0), packedY_(0), packedPhi_(0), packedM_(0), 
+    p4_(nullptr), p4c_(nullptr), vertex_(0,0,0),  pdgId_(0), charge_(0) { }
+    explicit PackedGenParticle( const reco::GenParticle & c)
+    : p4_(new PolarLorentzVector(c.pt(), c.eta(), c.phi(), c.mass())), p4c_( new LorentzVector(*p4_)), vertex_(0,0,0), pdgId_(c.pdgId()), charge_(c.charge()), mother_(c.motherRef(0)),
       statusFlags_(c.statusFlags()) { pack(); }
-  explicit PackedGenParticle( const reco::GenParticle & c, const edm::Ref<reco::GenParticleCollection> &  mother)
-    : p4_(c.pt(), c.eta(), c.phi(), c.mass()), p4c_(p4_), vertex_(0,0,0), pdgId_(c.pdgId()), charge_(c.charge()), mother_(mother), unpacked_(true),
```
Comment (Dr15Jones) [inline DataFormats/PatCandidates/interface/PackedGenParticle.h]: NOTE: the vertex is not taken from the GenParticle. Is this intended?

**Comment by Dr15Jones** [comment]
please test

**Comment by wmtan** [comment]
@davidlange6 @Dr15Jones Chris asked me to look at this, but I have no idea how to reproduce the problem. I do not know which workflows reproduce the problem. nor how to generate and view the distributions that do not match.

**Comment by Dr15Jones** [comment]
@wmtan I just drilled through the comparison to baseline plots and found 	50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50 to have a big difference if one clicks on that link, then goes to the 'ParticleFlow' and click on that pie chart then click on the pie chart for packedGenParticlesValidation then click on the 'CompWithGenParticles' pie chart.

So I suggest running workflow 50202.0. I searched from CompWithGenParticles and hit
http://cmslxr.fnal.gov/source/Validation/RecoParticleFlow/python/miniAODValidation_cff.py#0014
So you could look at that module to see what differences it is looking at.

**Comment by wmtan** [comment]
@Dr15Jones The problem easily reproduces. In step3 of relval 50202.0, pat::PackedGenParticle is produced. The distributions of all data members (e.g. pt, Y, etc) are totally screwed up. The differences are far from subtle, and you can tell it is screwed up just by looking. No need to compare with the distributions prior to this PR to tell the distributions are wrong. I'll try to debug this tomorrow.
Some good news: The pat::PackedCandidate distributions look OK, so the problem is confined to
pat::PackedGenParticle.

**Comment by davidlange6** [comment]
and this may or may not be related - but lots of warnings coming (I think) from these changes

https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc493/CMSSW_7_6_X_2015-10-19-1100/DataFormats/PatCandidates

**Comment by Dr15Jones** [comment]
We can make the warnings go away by using the macro CMS_THREAD_GUARD instead if the attribute directly.

**Comment by wmtan** [comment]
#12000 fixes the discrepancies. However, I did #12000 before I saw the comment about the thread guard warnings, so it does not fix those. The warnings are unrelated to the discrepancy.

**Pull request merged**